drivers/i2c/i2c_imx.c

```diff
@@ -238,8 +238,10 @@ static int i2c_imx_transfer(const struct device *dev, struct i2c_msg *msgs,
 		if ((msgs->flags & I2C_MSG_RW_MASK) == I2C_MSG_READ) {
 			i2c_imx_read(dev, msgs->buf, msgs->len);
 		} else {
-			if (!i2c_imx_write(dev, msgs->buf, msgs->len)) {
-				goto finish; /* No ACK received */
+			if (msgs->len) {
```

Comment:
Apparently I don't have this SoC, but #27896 fixed a similar problem where the NAK from the address did not get delivered promptly: it would normally arrive during the transfer of the data.  There the solution was to wait for some period to see whether it would show up.  Is that the problem seen on this platform?  If so, I don't see how this can work, as it simply bypasses the write and returns immediately with no error.

What platform are you using?

Reply:
Well it doesn't seem to be the same problem: without checking message length in this case `i2c_imx_write()` is called directly but that function doesn't check size of message before doing this:
https://github.com/zephyrproject-rtos/zephyr/blob/721b1f7c0c5cb901f1d49caafbe2e161ae2f62c2/drivers/i2c/i2c_imx.c#L69-L71
and so when a IRQ is received `transfer->txSize` is `0xff` and the buffer is somewhere in the address space

Reply:
sorry my platform is a custom board based on imx6sx and i've also a imx6sx sabre and a a pico pi imx7d

Reply:
I'm not questioning the need to avoid the write when the length is zero.  What I'm concerned about is what happens if an ack was received for a zero-byte transfer.

If this I2C peripheral guarantees to indicate the NAK immediately this fix is fine.  If, like the LPI2C, the NAK can be delayed, then it might falsely indicate presence of a device when one isn't actually there.

Absent evidence that a delay is possible I suspect this is fine, but want @MaureenHelm to weigh in.

Reply:
Ok now i understand :smile: , it's seem that the NAK is received immediately but to be honest i'll wait a confirmation from @MaureenHelm thanks